### PR TITLE
feat(Collective Tags): Autocompleting tag input & edit tags modal

### PR DESCRIPTION
**Pull request by gustavlrsn**
Resolve https://github.com/opencollective/opencollective/issues/5523
Require https://github.com/opencollective/opencollective-api/pull/8422

# Description

- A new autocompleting tags input, for Collective tags only (for now), with ability to reorder tags.
  - It can be supplied with `suggestedTags` that are shown below.
  - It's used in the create collective form, and added to the host application forms, where it is showing popular tags for that specific host, as well as in a new edit tags modal:
- `EditTagsModal` accessible from the collective page

# Screenrecordings/screenshots

https://user-images.githubusercontent.com/1552194/214555241-28f4dfdc-02dc-4698-8ca1-5110f382ae30.mov

image
image

### CODE REVIEW

**Comment by gustavlrsn** [comment]
Putting this in draft to fix a few more things:
- [ ] add this input to Form builder to enable it when editing a Collective
- [ ] add ability to reorder tags ([possible with react-select](https://react-select.com/advanced#sortable-multiselect))

**Comment by gustavlrsn** [comment]
Added an option to provide popular/suggested tags, see screenrecording in PR description.

**Comment by gustavlrsn** [comment]
Putting this in draft to add the Edit tags modal to this PR.

**Comment by Betree** [comment]
@gustavlrsn Let me know when this is ready for review

**Comment by gustavlrsn** [comment]
The "mutualize edit tags modal" commit should have been a stash, reverted now. 

It's a WIP to use the Edit tags modal for expenses, but [realized a UX issue with it explained in this Loom](https://www.loom.com/share/aefd030aa0d24fc2adf3fcfee93cb25e), so holding off on that for now.

File: package.json
Comment (Betree) [inline package.json:100]: It seems the use case for this lib was already covered by `react-dnd`, even though `react-sortable-hoc` seems to have a more straightforward API. Having two libs to achieve the same thing means more maintenance work and a lack of clarity for developers about which ones they should use.

If you think `react-sortable-hoc` is superior, could you open a follow-up issue about migrating all the other places (`AdminContributeCardsContainer`, `SocialLinks` and `EditCollectivePage`) to this new lib?

File: components/EditTagsModal.tsx
Comment (Betree) [inline components/EditTagsModal.tsx:29]: Is there a GraphQL V2 mutation we could use to achieve that?

File: components/EditTagsModal.tsx
Comment (Betree) [inline components/EditTagsModal.tsx:102]: I wasn't aware that Formik could focus on an arbitrary element. Could you explain what the bug was looking like?

File: components/EditTagsModal.tsx
Comment (Betree) [inline components/EditTagsModal.tsx:142]: No need to set `disabled` on loading buttons, clicks are prevented by default

```suggestion
                  disabled={!formik.dirty}
```

File: components/EditTagsModal.tsx
Comment (Betree) [inline components/EditTagsModal.tsx:150]: The widespread practice in the codebase is to mark only the save button as `loading`. I personally prefer this approach, having two loading buttons look strange to me.

```suggestion
```

image

File: components/StyledCheckbox.js
Comment (Betree) [inline components/StyledCheckbox.js:66]: Why remove this? Did you test to make sure there was no side-effect?

File: components/create-collective/index.js
Comment (Betree) [inline components/create-collective/index.js:169]: Nice :+1:

File: components/EditTagsModal.tsx
Comment (gustavlrsn) [reply components/EditTagsModal.tsx:150]: Ah, yes, this was unintentional, must have missed removing this when copying the Save button.

File: package.json
Comment (gustavlrsn) [reply package.json]: Yes, good point. I was a bit stuck in trying to make it work with the async version of react-select that accepts a `loadOptions` function, which did not play nicely with the `lodash/debounce` for some reason.

But I now took another look at the CollectivePicker and refactored the tags input to something similar, not using the async version and instead keep track of options and input state in the component.

Overall a much nicer solution than resorting to adding another library, thank you!

File: components/EditTagsModal.tsx
Comment (gustavlrsn) [reply components/EditTagsModal.tsx:29]: There is `editAccount` in V2 but it does not currently have tags editing implemented.

File: components/EditTagsModal.tsx
Comment (gustavlrsn) [reply components/EditTagsModal.tsx:102]: The autocomplete input was automatically opened when opening the modal, covering up the "Popular tags" options:
image

File: components/StyledCheckbox.js
Comment (gustavlrsn) [reply components/StyledCheckbox.js:66]: Added this back, a later fix that was successfully targeting the autocomplete dropdown to be on top fixed it. Wrong approach to mess with this here. Thanks for noticing!

File: package.json
Comment (gustavlrsn) [reply package.json]: Good point. Used this since it was easy to integrate with React-Select with an existing example, but agree that having two libraries for the same purpose is not great.

Had a look to compare what seems to be the better option overall, and found out that `react-sortable-hoc` is deprecated, so not a great option.

Tried adding `react-dnd` instead, but quickly found issues in making it work smoothly (apart from a more complicated API), such as items sometimes jumping back before they finally land on the correct destination, which is also an issue in the current places where it is being used.

Had a look at the library that `react-sortable-hoc` is being replaced by:[dnd-kit](https://github.com/clauderic/dnd-kit), and this seems like a better option with a simpler API and also seems more future-proof. 

The React-Select example is being updated to use `dnd-kit` as well: https://github.com/JedWatson/react-select/pull/5212

Added it in this PR and will open an issue to also migrate the existing uses of `react-dnd`.

**Comment by gustavlrsn** [comment]
> * When typing fast _tagName_ then pressing enter, the tag is not added and it tries to submit the form (using Firefox).

Fixed this!

File: package.json
Comment (gustavlrsn) [reply package.json]: Full replacement of `react-dnd` is in this PR: https://github.com/opencollective/opencollective-frontend/pull/8633

File: components/EditTagsModal.tsx
Comment (znarf) [reply components/EditTagsModal.tsx:29]: Would be great to support this in v2 indeed.

Comment (Betree) [reply components/EditTagsModal.tsx:29]: Doesn't have to be in this PR, but would be great to add it to `editAccount` or to use the new `setTags` mutation introduced in https://github.com/opencollective/opencollective-api/pull/8558

File: components/EditTagsModal.tsx
Comment (Betree) [reply components/EditTagsModal.tsx:102]: I see; I suppose `focus-trap-react` is the one triggering the focus then. Your approach to fix this looks right, I'd just clarify this in the comments:

```suggestion
            {/* A hack to disable automatic focus from focus-trap-react on the tags input */}
            {/* eslint-disable-next-line jsx-a11y/no-noninteractive-tabindex */}
```

File: package.json
Comment (Betree) [inline package.json:17]: For anyone wondering, bundlephobia returns 13.9kB + 3.7kB + 1.6kB for these, pretty much the same as `react-dnd`.

**Review by Betree — APPROVED** [review]
No blocker for me overall! Really cool implementation

**Pull request merged**
